keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,823 @@
+So, we limit the size of `succeededIndexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// SuccessPolicy specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// Specified it must be immutable and works only for the Indexed jobs.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	// +listType=atomic
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some indexes.
+type SuccessPolicy struct {
+	// Criteria represents the list of alternative criteria for declaring the jobs 
+	// as successful before its completion. Once any of the criteria are met,
+	// the "SuccessCriteriaMet" condition is added, and the lingering pods are removed.
+	// The terminal state for such a job has the "Complete" condition.
+	// Additionally, these criteria are evaluated in order; Once the job meets one of the criteria,
+	// other criteria are ignored.
+	//
+	// +optional
+	Criteria []SuccessPolicyCriterion
+}
+
+// SuccessPolicyCriteria describes a criterion for declaring a Job as succeeded.
+// Each criterion must have at least one of the "succeededIndexes" or "succeededCount" specified.
+type SuccessPolicyCriterion struct {
+	// SuccessPolicyCriterion specifies the set of indexes 
+	// which need to be contained in the actual set of succeeded indexes for the job.
+	// The list of indexes must come within 0 to ".spec.completions" and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	SucceededIndexes *string
+
+	// SucceededCount specifies the minimal size of the actual set of the succeeded indexes
+	// for the job.
+	// When specified it needs to be positive integer.
+	//
+	// +optional
+	SucceededCount *int
+	
+	// SuccessCriterionMetReasonSuffix specifies the CamelCase suffix of the reason for the "SuccessCriteriaMet" condition.
+	// Once the job meets this criterion, "JobSuccessPolicy" and the specified suffix is combined with "As".
+	// For example, if specified suffix is "LeaderSucceeded", it is represented as "JobSuccessPolicyAsLeaderSucceeded".
```

Comment:
I think putting this in the message is better than in the Reason so the user has more flexibility, but still gets the information they expect out of it.

Reply:
For end-users, it makes sense, but the Reason seems to be useful when we specify the criteria to mark CustomJob (third part CRDs) succeeded (ex. JobSet) via Job API.

Here is a JobsSet FailurePolicy use case: https://github.com/kubernetes-sigs/jobset/pull/381/files#r1460001639

@deads2k WDYT?

Reply:
Reasons are encouraged to be machine readable and have relatively tight validation compared to messages.  (see https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go#L1523)

While there is some flexibility, allowing the user to define their own details about success or failure seems better suited `Message`

Reply:
It sounds much more reasonable. I will replace `SuccessCriterionMetReasonSuffix` with `SuccessCriterionMetAdditionlMessage`. Thank you!

Reply:
-1 to message
The point is to make this machine readable.

Reply:
cc: @danielvegamyhre @alculquicondor

Reply:
Yeah, please don't set a message.

Reply:
@deads2k @alculquicondor @danielvegamyhre I updated the story 3: https://github.com/kubernetes/enhancements/pull/4062/commits/44387936f1417cda11e92e9d5c0634099e667c9d

Reply:
With the update to use this as the entire reason, it's no longer merging field responsibility.  I'm not confident that someone could author a re-useable problem based a value that will vary between jobs it is consuming, but the API and ownership are clear.

Reply:
I actually slightly prefer to let the users specify only the suffix for the reason. It will support the use case for machine matching. Yet, the reason will clearly indicate it was due to SucceessPolicy. IMO, this can be handy if a job is debugged by someone else than the user who supplied the reasons.  

Anyway, it would be good to align the approach with https://github.com/kubernetes/enhancements/pull/4479.

Reply:
> IMO, this can be handy if a job is debugged by someone else than the user who supplied the reasons.

I agree that the suffix would be helpful, but it seems that we don't allow the reason field to have merging field responsibility. If we truly want to have suffixes, it seems that the message would be better since the message is more flexible rather than reason, but since we want to give configurable in the machine-readable field, not human-readable field, we shouldn't' give configurable to the message field.

It means that the current approach seems (the entire reason) would be better.

Reply:
> With the update to use this as the entire reason, it's no longer merging field responsibility. 

Gotcha, indeed adding the suffix the have two responsibilities in the field: being human and machine readable. However, one can say the field already merges these two responsibilities. Having said that, I'm ok with either way.

Reply:
We will discuss `SetSuccessCriteriaMetReason` when moving to the second alpha stage.

So, I'm closing this now.